SmartTutor/ContentModel/LearningObjects/ChallengeModel/FulfillmentStrategy/NameChecker/BasicNameChecker.cs

```diff
@@ -0,0 +1,120 @@
+﻿using RepositoryCompiler.CodeModel.CaDETModel.CodeItems;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations.Schema;
+using System.Linq;
+
+namespace SmartTutor.ContentModel.LearningObjects.ChallengeModel.FulfillmentStrategy.NameChecker
+{
+    [Table("BasicNameCheckers")]
+    public class BasicNameChecker : ChallengeFulfillmentStrategy
+    {
+        public List<NamingRule> NamingRules { get; set; }
+
+        public override HintDirectory EvaluateSubmission(List<CaDETClass> solutionAttempt)
+        {
+            var challengeHints = GetApplicableHintsForClassNames(solutionAttempt);
+            challengeHints.MergeHints(GetApplicableHintsForFieldNames(solutionAttempt));
+            challengeHints.MergeHints(GetApplicableHintsForMemberNames(solutionAttempt));
+            challengeHints.MergeHints(GetApplicableHintsForVariableNames(solutionAttempt));
+            challengeHints.MergeHints(GetApplicableHintsForParameterNames(solutionAttempt));
+            challengeHints.MergeHints(GetApplicableHintsForRequiredWords(solutionAttempt));
+            return challengeHints;
+        }
+
+        public override List<ChallengeHint> GetAllHints()
+        {
+            var challengeHints = new List<ChallengeHint>();
+            challengeHints.AddRange(NamingRules.Select(nr => nr.Hint));
+            return challengeHints;
+        }
+
+        private HintDirectory GetApplicableHintsForClassNames(List<CaDETClass> solutionAttempt)
```

Comment:
The following five functions are duplicate code. You should extract a method and give it as input the name for validation and the snippet id (FullName for class, signature for method, class FullName for field, method signature for variable and parameter of method).

We don't want our hints to be too obvious - You are using meaningless words somewhere in method X makes the trainee thing more than "this variable that has 2 words has a meaningless word".

Regarding extracting the name and identifier, try: 
var fields = GetFieldsFromClasses(solutionAttempt);
EvaluateNames(fields.Select(f => f.Parent.FullName), fields.Select(f => f.Name)) //new method

Since you can't access the parent for variables and parameters through your helper function, you'll need to remove these functions and inline their logic here.

Reply:
Noted. I have created a new method for evaluation of name. This will remove duplicate code.